src/main/java/org/spongepowered/common/data/processor/data/tileentity/FurnaceDataProcessor.java

```diff
@@ -0,0 +1,137 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.processor.data.tileentity;
+
+import com.google.common.collect.Maps;
+
+import net.minecraft.tileentity.TileEntityFurnace;
+
+import org.spongepowered.api.block.BlockTypes;
+import org.spongepowered.api.data.DataContainer;
+import org.spongepowered.api.data.DataHolder;
+import org.spongepowered.api.data.DataTransactionBuilder;
+import org.spongepowered.api.data.DataTransactionResult;
+import org.spongepowered.api.data.key.Key;
+import org.spongepowered.api.data.key.Keys;
+import org.spongepowered.api.data.manipulator.immutable.tileentity.ImmutableFurnaceData;
+import org.spongepowered.api.data.manipulator.mutable.tileentity.FurnaceData;
+import org.spongepowered.api.world.World;
+import org.spongepowered.common.data.manipulator.mutable.tileentity.SpongeFurnaceData;
+import org.spongepowered.common.data.processor.common.AbstractTileEntityDataProcessor;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class FurnaceDataProcessor extends AbstractTileEntityDataProcessor<TileEntityFurnace, FurnaceData, ImmutableFurnaceData> {
+
+    public FurnaceDataProcessor() {
+        super(TileEntityFurnace.class);
+    }
+
+    @Override
+    protected boolean doesDataExist(TileEntityFurnace tileEntity) {
+        return true; //if its an TileEntityFurnace it always has the data
+    }
+
+    @Override
+    protected boolean set(TileEntityFurnace tileEntity, Map<Key<?>, Object> keyValues) {
+        // getField() / setField()
+        //
+        // 0 : furnaceBurnTime -> NOT equal to passedBurnTime --> the remaining burn time
+        // 1 : currentItemBurnTime -> equal to maxBurnTime
+        // 2 : cookTime -> equal to passedCookTime
+        // 3 : totalCookTime -> equal to maxCookTime
+
+        final int passedBurnTime = (Integer) keyValues.get(Keys.PASSED_BURN_TIME); //time (int) the fuel item already burned
+        final int maxBurnTime = (Integer) keyValues.get(Keys.MAX_BURN_TIME); //time (int) the fuel can burn until its depleted
+        final int passedCookTime = (Integer) keyValues.get(Keys.PASSED_COOK_TIME); //time (int) the item already cooked
+        final int maxCookTime = (Integer) keyValues.get(Keys.MAX_COOK_TIME); //time (int) the item have to cook
+        final boolean needsUpdate = !tileEntity.isBurning() && maxBurnTime > 0;
+
```

Comment:
You'll need to validate that `passedBurnTime` and `passedCookTime` are not greater than their respective maximums (`maxBurnTime` and `maxCookTime`)